tools/config_generator/config_generator.cpp

```diff
@@ -0,0 +1,811 @@
+
+        if(std::holds_alternative<size_t>(is_two_phase_mode)
+           && std::get<size_t>(is_two_phase_mode) == 1) {
+            const auto coordinator_count
+                = get_param_from_template_file(coordinator_count_key,
+                                               config_map);
+            return_msg += create_component(sentinel_count_key,
+                                           std::get<size_t>(sentinel_count),
+                                           true);
+            return_msg += create_component(shard_count_key,
+                                           std::get<size_t>(shard_count),
+                                           true);
+            return_msg += create_component(coordinator_count_key,
+                                           std::get<size_t>(coordinator_count),
+                                           true);
+        } else {
+            const auto atomizer_count
+                = get_param_from_template_file(atomizer_count_key, config_map);
+            const auto archiver_count
+                = get_param_from_template_file(archiver_count_key, config_map);
+            const auto watchtower_count
+                = get_param_from_template_file(watchtower_count_key,
+                                               config_map);
+            return_msg += create_component(shard_count_key,
+                                           std::get<size_t>(shard_count),
+                                           false);
+            return_msg += create_component(sentinel_count_key,
+                                           std::get<size_t>(sentinel_count),
+                                           false);
+            return_msg += create_component(archiver_count_key,
+                                           std::get<size_t>(archiver_count),
+                                           false);
+            return_msg += create_component(atomizer_count_key,
+                                           std::get<size_t>(atomizer_count),
+                                           false);
+            return_msg += create_component(watchtower_count_key,
+                                           std::get<size_t>(watchtower_count),
+                                           false);
+        }
+        // add support for pre-seeding
+        // only has effect if pre-seeding is active
+        const auto seed_begin
+            = get_param_from_template_file(seed_from, config_map);
+        const auto seed_end
+            = get_param_from_template_file(seed_to, config_map);
+        if(!std::holds_alternative<std::string>(seed_begin)
+           && !std::holds_alternative<std::string>(seed_end)
+           && std::get<size_t>(seed_begin) != std::get<size_t>(seed_end)) {
+            std::pair<std::string, std::string> key_pair = create_key_pair();
+            set_param_to_config_file(seed_privkey, key_pair.first);
+            write_generated_config_to_file();
```

Comment:
Looks like this exhibits undesirable behavior; namely, it still writes the config to-file, and it seems to be triggered if `seed_{begin,end}` aren't specified in the template. Here's my local test session reproducing this:

```console
$ ./scripts/build.sh
<snip>
$ cd ./build/tools/config_generator
$ ./generate_config ../../../config/unit/2pc_template_unit_test.tmpl 5000
Warning: Could not find param, seed_from.
Warning: Could not find param, seed_to.
SUCCESS

```

Since the aim is for the generator to only produce configs on `stdout`, maybe we should completely delete `write_generated_config_to_file()`?

Reply:
This is a good note -- I neglected that this tries to print to stdout whereas the config file is printed to stderr when other paths are taken

Reply:
Sorry, just to clarify: the config should be written to `stdout`, the rest of the messages (like, `SUCCESS`) should be written to `stderr`; but the bigger thing is that in the path above, the config actually was never written to any visible stream.

Reply:
Got it, I will make that adjustment. It looked to me like previously `write_config_to_file()` was trying to write the value of `m_new_config` to stdout (possibly was uninitialized which is why nothing was printed, or that print was trapped).